Extended TSql Parser and Lexer

ealight opened the pull request:
- Added all the math functions to built-ins. ([Source](https://learn.microsoft.com/en-us/sql/t-sql/functions/mathematical-functions-transact-sql?view=sql-server-ver16))
- Extended date built-ins functions up-to-date. ([Source](https://learn.microsoft.com/en-us/sql/t-sql/functions/date-and-time-data-types-and-functions-transact-sql?view=sql-server-ver16))
- Added procedure's temporary tables recognizing. Now temporary tables which start with '#' are going to temp_table_name/full_temp_table_name contexts. ([Source](https://www.red-gate.com/simple-talk/databases/sql-server/t-sql-programming-sql-server/temporary-tables-in-sql-server/))
- Added database engine stored procedures handling. For now it's only 'sp_executesql' but there is a field to easy work with. ([Source](https://learn.microsoft.com/en-us/sql/relational-databases/system-stored-procedures/database-engine-stored-procedures-transact-sql?view=sql-server-ver16)).
- Added old fashion join - NON ANSI Join handling. They're going to own context. ([Source](https://sqlenlight.com/support/help/sa0006/))
- Added dateparts for date built-in functions to match documentation. (dateparts_9 - [DATE_BUCKET](https://learn.microsoft.com/en-us/sql/t-sql/functions/date-bucket-transact-sql?view=sql-server-ver16), dateparts_12 - [DATE_ADD](https://learn.microsoft.com/en-us/sql/t-sql/functions/dateadd-transact-sql?view=sql-server-ver16), [DATE_DIFF](https://learn.microsoft.com/en-us/sql/t-sql/functions/datediff-transact-sql?view=sql-server-ver16), [DATEDIFF_BIG](https://learn.microsoft.com/en-us/sql/t-sql/functions/datediff-big-transact-sql?view=sql-server-ver16), dateparts_15 - [DATENAME](https://learn.microsoft.com/en-us/sql/t-sql/functions/datename-transact-sql?view=sql-server-ver16), [DATEPART](https://learn.microsoft.com/en-us/sql/t-sql/functions/datepart-transact-sql?view=sql-server-ver16), dateparts_datetrunc - [DATETRUNC](https://learn.microsoft.com/en-us/sql/t-sql/functions/datetrunc-transact-sql?view=sql-server-ver16))

KvanTTT commented on sql/tsql/TSqlParser.g4:
```diff
@@ -4386,17 +4491,41 @@ entity_name_for_parallel_dw
 
+// https://www.red-gate.com/simple-talk/databases/sql-server/t-sql-programming-sql-server/temporary-tables-in-sql-server/
+temp_table_name
+    : (database=id_ '.' (schema=id_)? '.' | schema=id_ '.')? table=TEMP_ID
+    | (database=id_ '.' (schema=id_)? '.' | schema=id_ '.')? blocking_hierarchy=BLOCKING_HIERARCHY
```
Please factor the common part:

```suggestion
    : (database=id_ '.' schema=id_? '.' | schema=id_ '.')? (table=TEMP_ID | blocking_hierarchy=BLOCKING_HIERARCHY)
```

KvanTTT commented on sql/tsql/TSqlParser.g4:
```diff
@@ -4386,17 +4491,41 @@ entity_name_for_parallel_dw
+temp_full_table_name
+    : (linkedServer=id_ '.' '.' schema=id_   '.'
+    |                       server=id_    '.' database=id_ '.'  schema=id_   '.'
+    |                                         database=id_ '.' (schema=id_)? '.'
```
Redudant parentheses:

```suggestion
    |                                         database=id_ '.' schema=id_? '.'
```

KvanTTT reviewed (changes requested):
Please cover your changes with some examples (`sql/tsql/examples`).

ealight commented:
@KvanTTT 

- Added dateparts for date built-in functions to match documentation. (dateparts_9 - [DATE_BUCKET](https://learn.microsoft.com/en-us/sql/t-sql/functions/date-bucket-transact-sql?view=sql-server-ver16), dateparts_12 - [DATE_ADD](https://learn.microsoft.com/en-us/sql/t-sql/functions/dateadd-transact-sql?view=sql-server-ver16), [DATE_DIFF](https://learn.microsoft.com/en-us/sql/t-sql/functions/datediff-transact-sql?view=sql-server-ver16), [DATEDIFF_BIG](https://learn.microsoft.com/en-us/sql/t-sql/functions/datediff-big-transact-sql?view=sql-server-ver16), dateparts_15 - [DATENAME](https://learn.microsoft.com/en-us/sql/t-sql/functions/datename-transact-sql?view=sql-server-ver16), [DATEPART](https://learn.microsoft.com/en-us/sql/t-sql/functions/datepart-transact-sql?view=sql-server-ver16), dateparts_datetrunc - [DATETRUNC](https://learn.microsoft.com/en-us/sql/t-sql/functions/datetrunc-transact-sql?view=sql-server-ver16))
- Added examples for each case.
- Fixed 'hints.sql.tree' to test tree correctly with new changes.

ealight commented:
- Fixed '[xml_data_type.sql.tree](https://github.com/antlr/grammars-v4/pull/2994/files#diff-e09bd0b5a704c4cc2c6ebcb68641a5c283856b6ded48f54e75038bcd2f5ad28b)' to test tree correctly with new changes.

KvanTTT commented on sql/tsql/TSqlParser.g4:
```diff
@@ -4115,6 +4220,43 @@ xml_data_type_methods
 
+// https://learn.microsoft.com/en-us/sql/t-sql/functions/date-bucket-transact-sql?view=sql-server-ver16
+dateparts_9
+    : (YEAR | YEAR_ABBR)
+    | (QUARTER | QUARTER_ABBR)
+    | (MONTH | MONTH_ABBR)
+    | (DAY | DAY_ABBR)
+    | (WEEK | WEEK_ABBR)
+    | (HOUR | HOUR_ABBR)
+    | (MINUTE | MINUTE_ABBR)
+    | (SECOND | SECOND_ABBR)
+    | (MILLISECOND | MILLISECOND_ABBR)
```
Parentheses are excess:

```suggestion
    : YEAR | YEAR_ABBR
    | QUARTER | QUARTER_ABBR
    | MONTH | MONTH_ABBR
    | DAY | DAY_ABBR
    | WEEK | WEEK_ABBR
    | HOUR | HOUR_ABBR
    | MINUTE | MINUTE_ABBR
    | SECOND | SECOND_ABBR
    | MILLISECOND | MILLISECOND_ABBR
```

ealight replied:
Sure, I wanted to highlight datepart with their abbreviations but to lightweight parser it's really worth to remove the parentheses.

KvanTTT commented on sql/tsql/TSqlParser.g4:
```diff
@@ -4115,6 +4220,43 @@ xml_data_type_methods
+    ;
+
+// https://learn.microsoft.com/en-us/sql/t-sql/functions/dateadd-transact-sql?view=sql-server-ver16
+dateparts_12
+    :
+    | dateparts_9
+    | (DAYOFYEAR | DAYOFYEAR_ABBR)
+    | (MICROSECOND | MICROSECOND_ABBR)
+    | (NANOSECOND | NANOSECOND_ABBR)
+    ;
```
Is it ok that `dateparts_12` has the empty alternative? If so, plese remove the empty alternative here and insert `?` to all `dateparts_12` usings. Also, parentheses are excess.

```suggestion
dateparts_12
    :
    | dateparts_9
    | DAYOFYEAR | DAYOFYEAR_ABBR
    | MICROSECOND | MICROSECOND_ABBR
    | NANOSECOND | NANOSECOND_ABBR
    ;
```

ealight replied:
Oh, empty alternative isn't okey here. This is my inattention. Parentheses also deleted.

KvanTTT commented on sql/tsql/TSqlParser.g4:
```diff
@@ -4115,6 +4220,43 @@ xml_data_type_methods
+    ;
+
+// https://learn.microsoft.com/en-us/sql/t-sql/functions/datename-transact-sql?view=sql-server-ver16
+dateparts_15
+    : dateparts_12
+    | (WEEKDAY | WEEKDAY_ABBR)
+    | (TZOFFSET | TZOFFSET_ABBR)
+    | (ISO_WEEK | ISO_WEEK_ABBR)
+    ;
```
```suggestion
dateparts_15
    : dateparts_12
    | WEEKDAY | WEEKDAY_ABBR
    | TZOFFSET | TZOFFSET_ABBR
    | ISO_WEEK | ISO_WEEK_ABBR
    ;
```

KvanTTT commented on sql/tsql/TSqlParser.g4:
```diff
@@ -4115,6 +4220,43 @@ xml_data_type_methods
+    ;
+
+// https://learn.microsoft.com/en-us/sql/t-sql/functions/datetrunc-transact-sql?view=sql-server-ver16
+dateparts_datetrunc
+    : dateparts_9
+    | (DAYOFYEAR | DAYOFYEAR_ABBR)
+    | (MICROSECOND | MICROSECOND_ABBR)
+    | (ISO_WEEK | ISO_WEEK_ABBR);
```
Excess parentheses + formatting (semicolon should be on new line):

```suggestion
dateparts_datetrunc
    : dateparts_9
    | DAYOFYEAR | DAYOFYEAR_ABBR
    | MICROSECOND | MICROSECOND_ABBR
    | ISO_WEEK | ISO_WEEK_ABBR
    ;
```

KvanTTT commented on sql/tsql/TSqlParser.g4:
```diff
@@ -4386,15 +4528,37 @@ entity_name_for_parallel_dw
+    ;
+
+persist_table_name
+    : (database=id_ '.' (schema=id_)? '.' | schema=id_ '.')? (table=id_ | blocking_hierarchy=BLOCKING_HIERARCHY)
```
```suggestion
    : (database=id_ '.' schema=id_? '.' | schema=id_ '.')? (table=id_ | blocking_hierarchy=BLOCKING_HIERARCHY)
```

KvanTTT commented on sql/tsql/TSqlParser.g4:
```diff
@@ -4386,15 +4528,37 @@ entity_name_for_parallel_dw
+
+// https://www.red-gate.com/simple-talk/databases/sql-server/t-sql-programming-sql-server/temporary-tables-in-sql-server/
+temp_table_name
+    : (database=id_ '.' (schema=id_)? '.' | schema=id_ '.')? (table=TEMP_ID | blocking_hierarchy=BLOCKING_HIERARCHY)
```
```suggestion
    : (database=id_ '.' schema=id_? '.' | schema=id_ '.')? (table=TEMP_ID | blocking_hierarchy=BLOCKING_HIERARCHY)
```

KvanTTT commented:
Welcome to the grammars repository and thanks for the contribution. But please fix remaining issues before merging.

ealight commented:
@KvanTTT, Done.

pguyot commented:
@ealight thank you for your contribution.

I am confused by the temp table rule changes. It makes rules `full_table_name` and `table_name` ambiguous which slows down parsing (about 5-10% on a quick benchmark). Indeed, `id_` can be a `TEMP_ID`.

Should we revert this?

ealight commented:
@pguyot hi, it would be great to leave distinction between temporary tables and persisted tables (according to documentation). If you have better solution without slowing down the parsing, you are welcome to revert current changes. Thanks.

pguyot commented:
My experience is that `temp_full_table_name` and `persist_full_table_name` are pretty useless because beyond ambiguity, temporary table names can be accepted by `persist_full_table_name` if they are quoted with brackets `[#table]`.

I'm not sure grammar can catch this and you need to parse the identifier eventually to distinguish temporary tables and persistent tables.

So I think we should revert this change.
